#162727418 User should reset password

daniellamarr opened the pull request:
#### What does this PR do?
- This PR allows users to reset their password

#### Description of Task to be completed?
- Create a screen to allow users to input their email address to get password reset links
- Create a screen for password reset
- Implement password reset logic

#### How should this be manually tested?
- Install dependencies with `npm i`
- Start the development server with `npm run start:dev`
- Navigate to `http://localhost:5050/forgot-password` in the browser to see email verification field
- Once a link has been sent, go to the mailbox and click on link to navigate to password reset screen.

#### Any background context you want to provide?
- The user must be signed up on the platform for the request to be completed

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/162727418

#### Screenshots (if appropriate)
<img width="1440" alt="screen shot 2019-02-15 at 1 24 42 pm" src="https://user-images.githubusercontent.com/42177767/52856609-312b8f00-3125-11e9-9170-6b8006b31acf.png">
<img width="1440" alt="screen shot 2019-02-15 at 1 24 51 pm" src="https://user-images.githubusercontent.com/42177767/52856611-312b8f00-3125-11e9-9538-ab315473c631.png">

#### Questions:
N/A

daniellamarr committed:
Feature(User should be able to reset password):
Install url-loader, file-loader to handle file import on frontend
Add forgot password screen
Add forgot password route
Refactor webpack

daniellamarr committed:
Feature(User can reset password):
Created routes to handle reset password
Refactor webpack
Setup redux action and reducers to handle password reset

daniellamarr committed:
User can reset password

daniellamarr committed:
Feature(User should be able to reset password):
Implement password reset logic
Refactor ForgotPassword screen
[Finishes #162727418]

daniellamarr committed:
Feature(User should be able to reset password):
Remove sendResetPasswordEmail function from Login component
[Finishes #162727418]

daniellamarr committed:
Feature(User should be able to reset password):
Remove sendResetPasswordEmail function from Login component
[Finishes #162727418]

daniellamarr committed:
User should reset password
Fix merge conflicts
[Fixes #162727418]

daniellamarr committed:
User should reset password
Remove console.log from codebase
[Fixes #162727418]

jhaastrup committed:
Feature(User should be able to reset password):
Write tests for actions, components and reducers

Commit:
Feature(User should reset password):
Resolve merge conflicts
[Fix #162727418]

andela-dbamidele commented:
@daniellamarr . What's up on this PR?

daniellamarr commented:
> @daniellamarr . What's up on this PR?

Wow, a whole lot of conflicts I would have to fix

Commit:
Fix merge conflicts

andela-dbamidele commented:
Can you please adjust the modal headers? It should be consistence with the login modal

andela-dbamidele commented on package.json:
```diff
@@ -69,6 +69,7 @@
     "postcss-loader": "^3.0.0",
     "precss": "^4.0.0",
     "prop-types": "^15.7.1",
+    "query-string": "^6.2.0",
```
Please don't use this package. Create a simple helper function that will do the same thing as this guy.

daniellamarr replied:
I have resolved this

andela-dbamidele commented on src/__tests__/actions/authActions.test.js:
```diff
@@ -73,9 +82,63 @@ it('creates SET_USER_ERROR', async () => {
+    type: RESET_PASSWORD_EMAIL,
+    response: resetEmailResponse.data,
+  }];
+  Axios.post(`${basePath}/api/v1/auth/forgot-password`, resetEmail)
```
why are you using axios? You should be  mocking this.

andela-dbamidele commented on src/__tests__/actions/authActions.test.js:
```diff
@@ -73,9 +82,63 @@ it('creates SET_USER_ERROR', async () => {
+  Axios.post(`${basePath}/api/v1/auth/forgot-password`, resetEmail)
+    .then(res => expect(resetEmailResponse.data.token).toBe(res.data.data.token));
+  store.dispatch(sendResetPasswordEmail(resetEmail));
+  expect(resetEmailResponse.data.token).toBeTruthy();
```
I actually don't understand this test at all. I'll need some explanations.

andela-dbamidele commented on src/__tests__/components/ForgotPassword.test.js:
```diff
@@ -0,0 +1,71 @@
+    expect(wrapper.find('form').length).toEqual(1);
+  });
+  it('should submit form', () => {
+    wrapper.find('form').simulate('submit', {
```
where is the assertion?

andela-dbamidele commented on src/__tests__/components/ForgotPassword.test.js:
```diff
@@ -0,0 +1,71 @@
+  it('should render', () => {
+    const props2 = { ...props, auth: { isLoading: true } };
+    wrapper = mount(<ForgotPassword {...props2}/>);
+    expect(wrapper.find('button')).toEqual({});
```
What are you trying to achieve here? Is there a better way to test this?

andela-dbamidele commented on src/assets/scss/forgotPassword.scss:
```diff
@@ -0,0 +1,178 @@
+$card-width: 350px;
+$card-input-width: calc(350px - 10px);
+
+#forgotPassword {
```
Please can you convert all the `px` in this file to `em`?

andela-dbamidele commented on src/assets/scss/main.scss:
```diff
@@ -45,9 +45,12 @@ a:hover {
 button{
   height: 45px;
 }
+<<<<<<< HEAD
```
Merge conflicts?

daniellamarr replied:
Fixed

andela-dbamidele commented on src/assets/scss/main.scss:
```diff
@@ -45,9 +45,12 @@ a:hover {
   top: 0;
   z-index: 99;
 }
+>>>>>>> 24d01b1db41883091341070c4cfabf8e6cd49ed1
```
Same here

daniellamarr replied:
Fixed

andela-dbamidele commented on src/components/CompletePasswordReset.jsx:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  checkToken = () => {
+    const strings = queryString.parse(this.props.location.search);
```
Try and rename this to be more descriptive.

andela-dbamidele commented on src/components/CompletePasswordReset.jsx:
```diff
@@ -0,0 +1,109 @@
+    const { token } = strings;
+    if (!token) {
+      Notify.notifyError('This link is invalid or broken');
+      this.props.history.push('/login');
```
With this, will the user see the notification above?

andela-dbamidele commented on src/components/CompletePasswordReset.jsx:
```diff
@@ -0,0 +1,109 @@
+              <div>
+                <form method="post" onSubmit={this.onSubmit}>
+                  <input
+                  type="password"
```
Please format this and others like this.

andela-dbamidele commented on src/components/CompletePasswordReset.jsx:
```diff
@@ -0,0 +1,109 @@
+                </form>
+              </div>
+            </div>
+            <div className="clear"></div>
```
What's the purpose of this?

andela-dbamidele commented on src/components/ForgotPassword.jsx:
```diff
@@ -0,0 +1,87 @@
+              <div>
+                <form method="post" onSubmit={this.onSubmit}>
+                  <input
+                  type="text"
```
Format this please.

andela-dbamidele commented on src/components/ForgotPassword.jsx:
```diff
@@ -0,0 +1,87 @@
+
+ForgotPassword.propTypes = {
+  sendResetPasswordEmail: PropTypes.func.isRequired,
+  history: PropTypes.object.isRequired,
```
Please desist from using `PropTypes.object` and `PropTypes.array`. Use `PropTypes.shape()` instead. And refactor everywhere you've done this.

andela-dbamidele commented on src/components/authentication/Login.jsx:
```diff
@@ -92,9 +93,9 @@ export class Login extends Component {
-                    <a href={'#'}
-                       className="text-center text-light text-poppins">
-                       Forgot Password?</a>
+                    <a href="/forgot-password"
```
Why anchor tag instead of `Link`?

andela-dbamidele reviewed (changes requested):
Some couple of change request. Looking good though.

abejide001 commented on src/__tests__/components/CompletePasswordReset.test.js:
```diff
@@ -0,0 +1,86 @@
+  //         error: '',
+  //       },
+  //     };
+  //     const state = mapStateToProps(mock);
```
remove the code comments

abejide001 reviewed (changes requested):
Left a comment

Commit:
Feature(Reset Password):
Apply TTL feedback
Refactor Login component to use Link instead of a for routing
Refactor component tests
Uninstall package query-string
Create getQueryStrings utility
[Fixes #162727418]

daniellamarr commented:
> Can you please adjust the modal headers? It should be consistence with the login modal

Done

Commit:
Feature(User should reset password)
Resolve merge conflicts
[Fix #162727418]
